### DISCUSSION TITLE
Setting optimal threadpool size

### DISCUSSION THREAD

**Discussion by kstanikviacbs** [libvips/libvips#3308] [Q&A] [answered]
Hi! As always - first - thanks for great work! 

Reaching out to clarify things after introduced threadpool refactor [8.14] since I'm a bit confused with it. I'm wondering what might be the optimal value for my use case. 

The app I work on serves multiple images per second with few operations performed per each, it is virtualised and containeraised. Current value set via `vips_concurrency_set` (BTW: is this API still valid way?) is 1 and performs pretty good, but wanted to check if there's any performance boost I might get from increasing it in this case. Any help and guidance will be appreciated! 

Thanks,
Konrad

**Comment by jcupitt** [comment, marked as answer]
Hi Konrad,

Yes the API still works. It now sets the maximum size of the threadpool. You could try turning it up and see if you get a speedup -- it depends on how much parallelism is available in your pipeline.

The new system tracks how many threads in each threapool are idle, and how many are doing useful work. If there are lots of idle threads, the pool will shrink (though no lower than 2 I think); if only a few threads are idle, it'll expand up to the concurrency limit. This really helps systems with large numbers of cores.

Probably (though it depends) you'll see the best performance if you don't change the concurrency limit and let libvips handle this for you.

**Reply by kstanikviacbs** [reply]
Understood, thanks for clarifying - I guess I need to run some benchmarks on my side then since sounds like it's really case specific. 

> Probably (though it depends) you'll see the best performance if you don't change the concurrency limit and let libvips handle this for you.

Do you mean here leaving it as is (so 1) or just not setting anything explicitly and relying on default here?

**Reply by jcupitt** [reply]
I wouldn't call `vips_set_concurrency()` unless I'd done quite a bit of benchmarking and was sure it needed changing.

**Reply by kstanikviacbs** [reply]
Got it, thank you. Closing the topic.

**Reply by kstanikviacbs** [reply]
@jcupitt Re-opening the topic just for a while to bring some benchmarking results as an input. 

Basically - getting rid of `vips_set_concurrency()` call and letting libvips to manage and optimise it in fact decreased processing time by about 15 ms (which is around 15% faster with my inputs), but at the same time I see some relevant increase of memory consumption. Based on how I understand it it's totally reasonable and expected, right?  

If yes then I guess it's up to me to decide which parameter should be a key factor and then empirically determine optimal value passed to `vips_set_concurrency` which will balance those 2, correct?
